### PR TITLE
T15856

### PR DESCRIPTION
**Pull request by joaquimrocha**
https://phabricator.endlessm.com/T15856

### CODE REVIEW

File: tools/libeosshellcontent.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+        return (strings_dict, langs)
+
+    def _get_app_metadata(self, app_id):
+        metadata_file = os.path.join(self._content_dir, 'content', 'Default',
+                                     'apps', 'content.json')
+        try:
+            all_metadata = json.load(open(metadata_file))
+        except FileNotFoundError:
+            return None
```
Comment (dbnicholson) [inline tools/libeosshellcontent.py]: Why is this exception being swallowed? I'd think we'd want the process to fail if content.json could not be found.

File: tools/eos-content-merge [hunk 2]
```diff
@@ -4,12 +4,14 @@
...
 import json
+from libeosshellcontent import ShellContent
 import os
 import re
 import shutil
 import sys
 
-CONTENT_FILE = '/usr/share/eos-shell-content/content.json'
+CONTENT_DIR = os.getcwd()
+CONTENT_FILE = CONTENT_DIR + '/content/Default/apps/content.json'
```
Comment (dbnicholson) [inline tools/eos-content-merge]: How does this work? It looks like you've changed this tool so that it only runs from an eos-shell-content checkout, but this is intended to run from the installed eos-shell-content-dev. Unless I'm missing something, this looks broken.

File: tools/libeosshellcontent.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        self._content_dir = content_dir
+        self._translations, self._langs = self._get_translations_dict()
+
+    def _get_translations_dict(self):
+        """Get translations dictionary and list of languages.
+
+        It reads the translations from a predefined file and returns a dictionary
+        with the translated strings and a list of the available languages.
+        """
+        po_dir = os.path.join(self._content_dir, 'po')
```
Comment (dbnicholson) [inline tools/libeosshellcontent.py]: I don't believe this directory is available when eos-shell-content-dev is installed.

File: tools/eos-content-merge [hunk 4]
```diff
@@ -68,7 +70,7 @@ class App(object):
 
     def merge_desktop_file(self, outpath, inplace=False):
         # Find the content desktop file
-        inpath = '/usr/share/eos-shell-content/bundle/desktops/' + \
+        inpath = os.getcwd() + '/bundle/desktops/' + \
```
Comment (dbnicholson) [inline tools/eos-content-merge]: Same.

File: tools/dh_eoscontent [hunk 5]
```diff
@@ -80,6 +80,30 @@ sub app_desktop_file {
...
 
+sub app_appdata_file {
+	my $app_id=shift;
+	my $prefix=shift;
+	my $tmp=shift;
+
+	# See if this package has the appdata file
+	my $appdata = $tmp . $prefix . '/share/appdata/' . $app_id
+	    . '.appdata.xml';
+	verbose_print("Checking for app-data file $desktop");
```
Comment (dbnicholson) [inline tools/dh_eoscontent]: `s/$desktop/$appdata/`

File: tools/libeosshellcontent.py (hunk 1 above)
Comment (joaquimrocha) [inline tools/libeosshellcontent.py]: This code was copied from *eos-build*. I can let it fail as you say.

File: tools/libeosshellcontent.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+            root = root.find('component')
+
+        # We do not want to display screenshots from 3rd parties when we have
+        # provided some from the Shell Content, so we need to remove them.
+        # The removal should be done in the app object but this cannot be
+        # accomplished until there is API to do it in libappstream-glib.
+        self._remove_unneeded_screenshots(root)
+
+        node = xml.dom.minidom.parseString(ET.tostring(root, encoding='utf-8'))
+        return node.toprettyxml(indent='  ')
```
Comment (joaquimrocha) [inline tools/libeosshellcontent.py]: This was just for having pretty XML (indented). I searched how to do that with ElementTree but this version (as opposed to lxml's) doesn't have a direct way to prettify the output, so many people were suggesting just using the minidom for it. It's ugly code but we'll get pretty XML for a small price.

File: tools/eos-content-merge (hunk 2 above)
Comment (joaquimrocha) [inline tools/eos-content-merge]: Hmm, sorry. I did this as a test but since I finished this task only a few days after it I completely forgot I had done it (and missed it when I added the patch). It was not my intention to propose this.

File: tools/libeosshellcontent.py (hunk 3 above)
Comment (joaquimrocha) [inline tools/libeosshellcontent.py]: Should I install the po dir as part of the *eos-shell-content-dev* then? Or is there a better solution for this?

Comment (dbnicholson) [inline tools/libeosshellcontent.py]: It looks like the catalogs are all installed in the eos-shell-content package, so you can use those. This requires a little adjustment. First, you'd glob all the `eos-shell-content.mo` files.
```
mofiles = glob.glob('/usr/share/locale/*/LC_MESSAGES/eos-shell-content.mo')
```
Then you'd loop over the files, detecting the language from the path and using the `mofile` function instead of `pofile`.
```
for f in mofiles:
    lang = f.split('/')[4]
    mo = polib.mofile(f)
    strings_dict[lang] = {(entry.msgid, entry.msgctxt): entry.msgstr for entry in mo}
```
The MOEntry class has the same attributes as POEntry. What you lose in the mo file is the untranslated strings, which I don't think we care about here.

**Comment by dbnicholson** [comment]
Another option is to generate the appdata.xml files in eos-shell-content like the desktop files are. Then write a tool to merge that with an existing appdata.xml file like we do with desktop files. Then we could also drop most of the code in eos-build that handles generating an appdata.xml file and instead just copy the appdata.xml file from eos-shell-content for bundling.

**Comment by joaquimrocha** [comment]
Thanks again @dbnicholson . I have decided to keep the Python module because it can still be useful even if run over the .mo files instead of the .po.
Please also verify if the path that I added to the *eos-shell-content-dev.install* is the correct one.

File: tools/Makefile.am [hunk 7]
```diff
@@ -3,6 +3,9 @@ EXTRA_DIST =
 CLEANFILES =
 DISTCLEANFILES =
 
+libeosshellcontentdir = $(pkgpythondir)
+libeosshellcontent_PYTHON = libeosshellcontent.py
```
Comment (dbnicholson) [inline tools/Makefile.am]: I think you missed my comments here before. Hmm, actually I don't see them so maybe I forgot to press a button or something.

* Putting this in `pkgpythondir` will be broken since that would make it `site-packages/eos-shell-content/libeosshellcontent.py`, which is invalid since `eos-shell-content` has `-`s in it. Just using `python_PYTHON =` will put it directly in `site-packages`, I believe.

* I don't think the `lib` prefix in `libeosshellcontent.py` is necessary. This is almost never done with python modules (look at all the other stuff you're importing). I think just `eosshellcontent.py` would be nicer.

File: tools/eos-content-merge (hunk 4 above)
Comment (dbnicholson) [inline tools/eos-content-merge]: This is still incorrect. Did you run this at all?

File: tools/libeosshellcontent.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        """Get translations dictionary and list of languages.
+
+        It reads the translations from a predefined file and returns a dictionary
+        with the translated strings and a list of the available languages.
+        """
+        strings_dict = {}
+        langs = []
+
+        for lang in os.listdir(LOCALES_DIR):
+            mo_file_path = LOCALES_DIR + lang + '/LC_MESSAGES/eos-shell-content.mo'
```
Comment (dbnicholson) [inline tools/libeosshellcontent.py]: `os.path.join`?

File: tools/eos-content-merge (hunk 4 above)
Comment (joaquimrocha) [inline tools/eos-content-merge]: I did run it but I forgot to include the change back (or rather undo the change) in the patch...

File: debian/control [hunk 9]
```diff
@@ -39,8 +39,10 @@ Depends: eos-shell-content (= ${binary:Version}),
          ${misc:Depends},
          ${perl:Depends},
          debhelper,
+         libappstream-glib,
```
Comment (dbnicholson) [inline debian/control]: There's actually no libappstream-glib package. What you want is gir1.2-appstreamglib-1.0.

File: tools/eos-content-merge [hunk 10]
```diff
@@ -136,6 +137,22 @@ class App(object):
             self.verbose_print('Updated desktop file follows:')
             outdesktop.write(sys.stdout, space_around_delimiters=False)
 
+    def merge_appdata_file(self, outpath, inplace=False):
+        shell_content = ShellContent()
+        xml = shell_content.update_appdata_from_file(outpath, self.appid)
+        if not xml:
+            self.verbose_print('Failed to get updated XML from ', outpath)
+            self.verbose_print('Not saving file!')
```
Comment (dbnicholson) [inline tools/eos-content-merge]: If the XML could not be read, shouldn't this error or at least return? The code below expects that the XML was read.

File: debian/control (hunk 9 above)
Comment (joaquimrocha) [inline debian/control]: Sorry, I always forget the *8* suffix that we have for some reason on Debian. On Fedora it's just *libappstream-glib*.
Packaging is so much fun!
